044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,173 @@
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: controller is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: controller is listening on the port of the first address in `controller.listener.names` (9090)
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: broker is listening on the port of the address of `inter.broker.listener.name` (9091)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the brokers will not become ready until a majority of the controllers 
+are up and running.
+This is similar to the current behaviour of brokers when ZooKeeper is not ready.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: node is listening on the port of the first address in `controller.listener.names` (9090)
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
+This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
+means the nodes will be able to communicate even if they are currently marked as not ready.
+
+### Impact on the KafkaRoller
+
+The existing KafkaRoller checks whether a pod is marked as ready to determine whether it needs to be rolled. 
+This relies on the fact that currently the pod being ready implies that the BrokerState is >= RUNNING.
+Additionally, it does not take into account the state of other broker pods.
+The move to KRaft mode introduces some new requirements in terms of when a pod should or should not be rolled.
+This combined with the varying readiness checks means we need to change the way the KafkaRoller works:
+
+* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.
+* If more than one controller pod has not become ready, KafkaRoller should try to role the controllers that aren't the current active controller first.
+* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to STARTED and are marked as ready
```

Comment:
is `STARTED` actually the `RUNNING` state?

Reply:
Yes good spot, I've updated it